crasm.1
.\" Copyright (C) 1987- Leon Bottou
.\"
.\" This is free documentation; you can redistribute it and/or
.\" modify it under the terms of the GNU General Public License as
.\" published by the Free Software Foundation; either version 2 of
.\" the License, or (at your option) any later version.
.\"
.\" The GNU General Public License's references to "object code"
.\" and "executables" are to be interpreted as the output of any
.\" document formatting or typesetting system, including
.\" intermediate and printed output.
.\"
.\" This manual is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public
.\" License along with this manual. Otherwise check the web site
.\" of the Free Software Foundation at http://www.fsf.org.
.\"
.TH CRASM 1 "1/9/1987" "Crasm-1.3" "Crasm-1.3"
.de SS
.SH \\0\\0\\0\\$*
..
.SH NAME
crasm \- Cross assembler for 6800/6801/6803/6502/65C02/Z80

.SH SYNOPSIS
.BI "crasm [-o " "codefile" "] [-slx] " "asmfile"

Assemble the microprocessor program 
.I asmfile 
and produce output file 
.I codefile
in Intel HEX or Motorola S Code format.
A program listing and a symbol table are also produced
on the standard output.
The current version of
.B crasm
can assemble programs for the 
.BR 6800, 
.BR 6801, 
.BR 6803, 
.BR 6502, 
.BR 65C02, 
and
.BR Z80
processors.
The full list is printed when you invoke
.B crasm
without arguments.

.SH OPTIONS
.TP
.BI "-o " codefile
Specify the name of the output file.
No output file is produced without this option.
.TP
.BI "-s"
Disable warnings.
.TP
.BI "-l"
Disable program listing output.
.TP
.BI "-x"
Disable symbol table output.
.PP


.SH SYNTAX

Each line of the assembly program
should follow one of the following templates,
where the brackets delimit optional parts.
.IP ""
.BI "[;" comment "]"
.br
.BI "" label " = " expression " [;" comment "]"
.br
.BI "[" label "] " mnemonic " " operand " [;" comment "]"
.br
.BI "" label "[;" comment "]"
.PP
Comments are introduced by a semicolon 
.BR "" "(" ";" ")"
and extend to the end of the line.
Labels are identifiers containing up to 36 alphanumeric 
characters (including period and underscore). Labels
cannot start with a digit.
Crasm does not usually pays attention to the column position
of the line components. The only exception is the case
of a line containing only a label and possibly a comment.
In that case the label must start in column zero.
.PP
The format of the mnemonics and operands 
field depends on the selected micro-processor.
A few mnemonics are valid for all processors
and are used to give directives to the assembled.
These are known as "pseudo-mnemonics".

.SS Labels

Labels are identifiers representing
.br
\(em an absolute address,
.br
\(em a relative address (position independent code),
.br
\(em a register,
.br
\(em a list of registers,
.br 
\(em a specific bit at a specific address,
.br
\(em or a mnemonic.
.PP
Most labels are composed of at most 36 alphanumeric characters,
periods
.BR "" "(" "." ")"
or
underscores
.BR "" "(" "_" ")."
Labels cannot start with a digit.  
They are case insensitive.
Labels starting with a period
.BR "" "(" "." ")"
are local labels whose scope is either limited
to the macro in which they are defined,
or to the code segment delimited by the pseudo-mnemonics
.B CODE
or
.BR DUMMY .
.PP
The predefined "star" label 
.BR "" "(" "*" ")"
represents the current program counter,
that is to say, the address where the next
assembly code instruction will be encoded.
Other predefined labels include all pseudo-mnemonics, 
micro-processor specific mnemonics and 
register names.


.SS Constants

The assembled recognizes numerical constants
expressed in decimal, hexadecimal, octal,
binary, or ascii.
.PP
.TS
center,box;
lfB lfB lfB
l l l 
l l l 
l l l 
l l l 
l l l 
l l l 
l l l 
l l l
l l l
.
Type	Format	Examples
=
decimal	\fIdddd	\fB1234\fR,\fB 675\fR,\fB 12\fR,\fB 1\fR, but not \fB0.12\fR.
=
hexadecimal	\fB$\fIdddd	\fB$fd12\fR,\fB $2AC\fR,\fB $0\fR.
	\fIdddd\fBH\fR	\fB03H\fR,\fB 2da7H\fR,\fB 0FC84H\fR, but not \fBFC84H\fR.
	\fB0X\fIdddd	\fB0x03\fR,\fB 0x2AC\fR,\fB 0Xfc84\fR.
=
octal	\fIdddd\fBQ\fR	\fB377Q\fR,\fB 012412Q\fR.
=
binary	\fB%\fIdddd	\fB%01110110\fR,\fB %1100\fR.
	\fIdddd\fBB\fR	\fB01110110B\fR,\fB 1100B\fR.
	\fB0B\fIdddd	\fB0b1100\fR
=
ascii	\fB'\fIcccc\fB'\fR	\fB'a'\fR,\fB 'AB'\fR,\fB '"'\fR,\fB '\\n'\fR,\fB '\\''\fR.
	\fB"\fIcccc\fB"\fR	\fB"\\t"\fR,\fB "\\""\fR,\fB "a'b"\fR.
.TE
.PP

.SS Expressions

Like labels, expressions can represent an absolute address
.RI ( abs ),
a relative address for position independent code
.RI ( rel ),
a register
.RI ( reg ),
or a list of registers
.RI ( reglist ),
or a reference to a specific bit at a specific address
.RI ( bspec ).

The following operators are recognized on expressions.
.TS
center,box;
lfB lfB lfB
l l l 
l l l 
l l l 
l l l 
l l l 
l l l 
l l l 
l l l 
l l l 
l l l 
l l l 
l l l 
l l l 
l l l 
l l l 
l l l 
.
Syntax	Result	Description
=
 \fIabs\fB{\fIabs\fB}\fR	\fIbspec\fR	bit reference, e.g. \fBpia{3}\fR
 \fBADDR(\fIabs\fB)\fR	\fIabs\fR	address from a bit reference
 \fBBIT(\fIabs\fB)\fR	\fIabs\fR	bit number from a bit reference
=
 \fB- \fIabs	\fIabs\fR	two's complement
 \fB~ \fIabs	\fIabs\fR	one's complement
=
 \fIabs\fB << \fIabs\fR	\fIabs\fR	left shift
 \fIabs\fB >> \fIabs\fR	\fIabs\fR	right shift
 \fBHI(\fIabs\fB)\fR	\fIabs\fR	high byte ((abs>>8)&$ff)
 \fBLO(\fIabs\fB)\fR	\fIabs\fR	low byte (abs&$ff)
=
 \fIabs\fB | \fIabs\fR	\fIabs\fR	bitwise or
 \fIabs\fB & \fIabs\fR	\fIabs\fR	bitwise and
 \fIabs\fB ^ \fIabs\fR	\fIabs\fR	bitwise xor
=
 \fIabs\fB * \fIabs\fR	\fIabs\fR	multiplication
 \fIabs\fB * \fIabs\fR	\fIabs\fR	division
=
 \fIabs\fB + \fIabs\fR	\fIabs\fR	addition
 \fIrel\fB + \fIabs\fR	\fIrel\fR	addition
 \fIabs\fB - \fIabs\fR	\fIabs\fR	subtraction
 \fIrel\fB - \fIabs\fR	\fIrel\fR	subtraction
 \fIrel\fB - \fIrel\fR	\fIabs\fR	subtraction
=
 \fIreg\fB - \fIreg\fR	\fIreglist\fR	register range
 \fIreglist\fB \\ \fIreg\fR	\fIreglist\fR	register list
.TE
.PP
The table lists operators in order of decreasing precedence.
Parenthesis can be used to avoid ambiguities.
A warning is generated when an entire expression is surrounded
with parenthesis and can be confused with a micro-processor addressing mode.
.PP
Examples:
.IP ""
\fB (base+$12) >> 8 & 0xff00 \fR
.br
\fB 'A'-80H \fR
.br
\fB (base+0x12) \fR
.PP
The last example causes a warning because the parenthesis
were not necessary and might suggest a micro-processor 
addressing mode.
.PP
All arithmetic expressions are evaluated on 32 bits.
Arithmetic operations overflow silently.
The arithmetic values are then truncated to the
size implied by the micro-processor mnemonic.
This truncation might cause a warning message.

Examples:
all the following instructions
.IP ""
(6502)	\fBlda #$1234\fR
.br
(6800)	\fBldaa $1234,x\fR
.br
(Z80)	\fBld (ix+0C2H),b\fR
.PP
cause a warning
.IP ""
.B >>> WARNING: Operand overflow
.PP
However expression
.IP ""
.B $1123454 * 1298992
.PP
overflows silently.

.SS Pseudo-mnemonics

The following pseudo-mnemonics are always recognized.

.PP
.BI "CPU " cpuname
.in +7n
Indicates the selected micro-processor type.
This must appear before anu micro-processor specific instruction.
The possible values of 
.I cpuname
are listed when you invoke
.B crasm
without arguments.
The current list includes
.BR 6800, 
.BR 6801, 
.BR 6803, 
.BR 6502, 
.BR 65C02, 
and
.BR Z80
.in -7n
.PP
.BI "OUTPUT " binformat
.in +7n
Indicates the format of the output file.
Argument 
.I binformat
can take values
.B SCODE
for producing an output file using Motorola's S code, or
.B HEX
for Intel's Hex format.
The default depends on the selected micro-processor.
.in -7n
.PP
.BI CODE
.in +7n
Delimit the scope of local labels
and introduce a program section.
.TP
.BI DUMMY
Delimit the scope of local labels
and introduce a fake program section
whose sole effect is to define labels
without generating code.
.in -7n
.PP
.BI "" label " EQU " expression 
.br
.BI "" label " = " expression
.in +7n
Define the value of label
.IR label .
Labels defined using these directives
can be redefined later in the program.
.in -7n
.PP
.BI "[" label "] DB " expression "[,...," expression "]"
.in +7n
Insert the specified data bytes (8 bits).
.in -7n
.PP
.BI "[" label "] DW " expression "[,...," expression "]"
.in +7n
Insert the specified data words (16 bits).
The byte ordering depends on the selected micro-processor.
.in -7n
.PP
.BI "[" label "] DL " expression "[,...," expression "]"
.in +7n
Insert the specified data longs (32 bits).
The byte ordering depends on the selected micro-processor.
.in -7n
.PP
.BI "[" label "] DDB " expression "[,...," expression "]"
.in +7n
Insert the specified double bytes (16 bits).
The byte ordering is the opposite of the usual
byte ordering for the selected micro-processor.
.in -7n
.PP
.BI "[" label "] ASC " stringconstant
.in +7n
Insert the ascii representation of the string
.I stringconstant .
The string must be delimited by double quotes.
The C escape sequences \fB\\r\fR, \fB\\n\fR, \fB\\t\fR, 
\fB\\0\fR, \fB\\'\fR, \fB\\"\fR, 
and \fB\\\\\fR are recognized.
.in -7n
.PP
.BI "[" label "] DS " countexpr ",[" valexpr "]"
.in +7n
Insere
.I countexpr
bytes with value
.IR valexpr .	
The default value is zero.
.in -7n
.PP
.BI "[" label "] ALIGN EVEN"
.br
.BI "[" label "] ALIGN ODD"
.in +7n
Insert a null byte in order to make the program counter even or odd.
.in -7n
.PP
.BI "IF [" expr1 " [" oper " " expr2 "]]"
.br
 ...
.br
.BI "ELSE"
.br
 ...
.br
.BI "ENDC"
.in +7n
Conditional assembly.
The argument of
.B IF
can be empty, can be a single expression
.BR expr1 ,
or can be two expressions separated by a comparison operator
.BR > ", " < ", " >= ", " <= ", " == ", " != ", " = ", or " <> "."
The assembler processes the lines located between the
.B IF
and the
.BR ELSE 
pseudo-mnemonics if the comparison
.I "expr1 oper expr2"
is true or if the single expression
.I expr1
is nonzero.
Otherwise it processes the lines located between the
.B ELSE
and the
.BR ENDC
pseudo-mnemonics.
Conditional assembly instructions can be nested. The 
.B ELSE 
part can be omitted.
.in -7n
.PP
.BI "" "label" " MACRO"
.br
        ...
.br
.BI "      ENDM"
.br
.in +7n
Define a new mnemonic 
.I label
equivalent to all the instructions located
between the 
.B MACRO
and
.B ENDM
pseudo-mnemonics.
Invocations of the macro can specify a list of comma 
separated operands.  The character sequences 
\fB\\1\fR, \fB\\2\fR, ... \fB\\\fIN\fR
in the macro definition are replaced by the supplied operands.
The character sequence \fB\\#\fR is replaced by the number
of supplied operands.
The character sequence \fB\\*\fR is replaced by the
unparsed macro arguments.
.in -7n
.PP
.BI "EXITM"
.in +7n
This pseudo mnemonic can be used inside a macro definition
to exit the macro.  This is useful in conjunction with
the conditional assembly pseudo-mnemonics.
.in -7n
.PP
.BI "INCLUDE " filename
.in +7n
Force the assembler to process file named
.I filename
at the current point.
.in -7n
.PP
.BR "LIST ON"
.br
.BR "LIST OFF"
.in +7n
Enable or disable the production of a listing
(default is on.)
.in -7n
.PP
.BR "CLIST ON"
.br
.BR "CLIST OFF"
.in +7n
Enable or disable the production of a listing
for the non active branches of a conditional assembly
construct (default is on.)
.in -7n
.PP
.BR "ILIST ON"
.br
.BR "ILIST OFF"
.in +7n
Enable or disable the production of a listing
for included files (default is off.)
.in -7n
.PP
.BR "MLIST ON"
.br
.BR "MLIST OFF"
.in +7n
Enable or disable the production of a listing
for the macro expansions (default is off.)
.in -7n
.PP
.BR "NAM " title
.in +7n
Provide name
.I title
for the header of the listing pages.
.in -7n
.PP
.BR "PAGE"
.in +7n
Start a new listing page.
.in -7n
.PP
.BR "PAGE " lines "," columns
.in +7n
Specify the size of a listing page.
.in -7n
.PP
.BR "SKIP " number
.in +7n
Skip
.I number
lines.
.in -7n
.PP
.BR "FAIL " message
.in +7n
Generate an error message
.IR message .
.in -7n
.PP

.SH EXAMPLE

.PP
Here is a small 6502 program:
.IP "" 2
\fB
     cpu 6502
.br
 cout = $fded ; display a character
.br
    * = $300  ; assemble at $300
.br
      code
.br
 pstring  ldy #0
.br
 .1       lda message,y
.br
          beq .2
.br
          jsr cout
.br
          iny          
.br
 .2       bne .1
.br
          rts
.br
 message  asc "This is the message\0"
.br
      code       
\fR
.PP    

.SH CREDITS

Leon Bottou, September 1987.



src/crasm.c
/*  Copyright (C) 1987- Leon Bottou
 *
 *  This is free documentation; you can redistribute it and/or
 *  modify it under the terms of the GNU General Public License as
 *  published by the Free Software Foundation; either version 2 of
 *  the License, or (at your option) any later version.
 *
 *  The GNU General Public License's references to "object code"
 *  and "executables" are to be interpreted as the output of any
 *  document formatting or typesetting system, including
 *  intermediate and printed output.
 *
 *  This manual is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public
 *  License along with this manual. Otherwise check the web site
 *  of the Free Software Foundation at http://www.fsf.org.
 */
/*
  CRASM: crasm.c
  noyau central
  LYB 9/87
*/

#include "dcl.h"
#include "version.h"
#include "label.h"

#define SCODE 1
#define LIST  2
#define XREF  4
#define WARN  8
#define INPUT 128

extern struct label* searchlabel();
extern struct label pseudos[];

jmp_buf errorjump;
int errnumber;
int warnnumber;
int segment;
extern int includelevel;
extern int thiscall;
struct label* starlabel;
int llen;
int ppos;
int plen;
int lpos;
int linenumber;
char lineprefix;
char* filename;
char* scodename;
char curline[256];

FILE* file;
FILE* scode;

int asmflags;
int passnumber;
int macrolevel;
int advance;

void jmp_buf_copy(jmp_buf out, jmp_buf in)
{
#ifdef JBUFSIZE
  int i;

  for (i = 0; i < JBUFSIZE; i++)
  {
    out[i] = in[i];
  }

#else
  memcpy(out, in, sizeof(jmp_buf));
#endif
}


int main(int argc, char** argv)
{
  int flags;
  char* s;

  passnumber = 0;
  flags = LIST | WARN;
#ifdef DEBUG
  flags |= XREF;
#endif

  if (argc == 0)
  {
    fatal("NOT from WB");
  }

  while (--argc)
  {
    if (**(++argv) == '-')
    {
      for (s = (*argv) + 1; *s; s++)
      {
        switch (*s)
        {
        case 's': /* suppress warnings */
          flags &= ~WARN;
          break;

        case 'l': /* suppress listing */
          flags &= ~LIST;
          break;

        case 'x': /* force XREF */
          flags |= XREF;
          break;

        case 'o': /* output name */
          if (!s[1] && argc > 1 && !(flags & SCODE))
          {
            flags |= SCODE;
            scodename = *++argv;
            --argc;
          }

          break;

        default:
          syntax("Bad Option(s)");
        }
      }
    }
    else
    {
      if (flags & INPUT)
      {
        syntax("One input only");
      }

      filename = *argv;
      flags |= INPUT;
    }
  }

  if (flags & INPUT)
  {
    crasm(flags);
  }
  else
  {
    syntax("No input!");
  }

  return 0;
}


void crasm(int flag)
{
  init_label_list(pseudos);
  file = fopen(filename, "r");

  if (file == NULL)
  {
    fatal("can't open source file");
  }

  scode = NULL;
  llen = 76;
  plen = 66;
  ppos = 0 ;
  lpos = 0;

#ifdef DEBUG
  asmflags = F_NOCODE;
#else
  asmflags = F_NOLIST | F_NOWARNING | F_NOERROR | F_NOCODE;
#endif

  pass(1);

#ifdef DEBUG
  outputEOP();
#endif
  asmflags &= ~F_NOERROR;

  if (flag & SCODE)
  {
    asmflags &= ~F_NOCODE;
  }

  if (flag & LIST)
  {
    asmflags &= ~F_NOLIST;
  }

  if (flag & WARN)
  {
    asmflags &= ~F_NOWARNING;
  }

  undeflabels(lroot);
  pass(2);

  if (scode)
  {
    closescodefile();
  }

  asmflags |= F_LIST_ON;

  if (errnumber == 0 && flag & LIST)
  {
    flag |= XREF;
  }

  result();

  if (errnumber && (flag & SCODE))
  {
    remove(scodename);
  }

  if (flag & XREF)
  {
    xref();
  }
}


void pass(int n)
{
  clearerr(file);
  rewind(file);

  asmflags |= F_CODE_ON;

  if (!(asmflags & F_NOLIST))
  {
    asmflags |= F_LIST_ON;
  }

#ifdef DEBUG
  asmflags |= F_IFLIST_ON | F_MACROLIST_ON;
#else
  asmflags &= ~ F_MACROLIST_ON;
  asmflags |=   F_IFLIST_ON;
#endif
  macrolevel    = 0;
  thiscall      = 0;
  segment       = 0;
  includelevel  = 0;
  linenumber    = 0;
  lineprefix    = 0;

  starlabel = findlabel("*");
  starlabel->flags |= UNDEF;
  starlabel->flags &= ~(NOREDEF | FORWARD);

  passnumber = n;
  fprintf(stderr, "Pass #%d\n", n);

  errnumber = warnnumber = 0;

  while (!feof(file))
  {
    if (setjmp(errorjump))
    {
      errnumber++;
    }

    linegets(curline, 250);

    if (asmline(curline, 3))
    {
      error("ELSE, ENDC, ENDM, EXITM illegal here");
    }
  }
}


static void herelabel(char* label)
{
  if (strcmp(label, "*") == 0)
  {
    error("illegal star definition");
  }

  if (asmflags & F_ORG_GV)
  {
    deflabel(label, NOREDEF,
             asmflags & F_RELATIF ? L_RELATIVE : L_ABSOLUTE,
             starlabel->value);
  }
  else
  {
    error("undefinable label: no org given");
  }

  if (!(asmflags & F_CODE_ON))
  {
    if (macrolevel == 0 || asmflags & F_MACROLIST_ON)
    {
      if (lpos == 0)
      {
        position(2);
        outputaddr(starlabel->value);
      }
    }
  }
}


static struct label* findmnemo(char* mnemo)
{
  char* aux1;
  char* aux2;
  register struct label* labmnemo;

  labmnemo = searchlabel(mnemo);

  if (labmnemo == NULL) /* sizer 68k */
  {
    if (filter(mnemo, "?.?", &aux1, &aux2))
    {
      if ((labmnemo = searchlabel(aux1)))
      {
        if (labmnemo->modifier >= 0)
        {
          labmnemo = NULL;
        }
      }
    }
  }

  if (labmnemo)
  {
    if (!(labmnemo->flags & UNDEF))
    {
      if (labmnemo->type == L_MNEMO ||
          labmnemo->type == L_MACRO)
      {
        return labmnemo;
      }
    }
  }

  return NULL;
}


static int nospacein(char* s)
{
  while (*s)
  {
    if (*s == ' ' || *s == '\t')
    {
      return FALSE;
    }

    s++;
  }

  return TRUE;
}


int asmline(char* s, int status)
{
  char* label;
  char* mnemo;
  char* oper;
  register struct label* labmnemo;
  int linestartswithoutblanks = 0;

  advance = 0;

  if (starlabel->flags & UNDEF)
  {
    asmflags &= ~F_ORG_GV;
  }
  else
  {
    asmflags |= F_ORG_GV;
    setpc(starlabel->value);
  }

  zerobuffer();

  if (*s != ' ' && *s != '\t')
  {
    linestartswithoutblanks = 1;
  }

  /* on traite
    XXX ; REMARK
     puis les cas
    XXX = blanc
  */
  if (filter(s, "_?_;?", &label, &mnemo))
  {
    s = label;
  }
  else
  {
    register char* p = s;

    while (*p)
    {
      p++;
    }

    while (p > s && (*--p == ' ' || *p == '\t'))
    {
      *p = 0;
    }

    while (*s == ' ' || *s == '\t')
    {
      s++;
    }
  }

  if (*s == 0)
  {
    if (status & 1)
    {
      outputline();
    }

    return 0;
  }

  /*   autorise:
      LABEL = MNEMO
    MNEMO
    LABEL MNEMO
      MNEMO OPERAND
    LABEL MNEMO OPERAND
       also:
    LABEL when line starts without blanks
  */

  if (filter(s, "?_=_?", &label, &oper) && nospacein(label))
  {
    mnemo = "equ";
    labmnemo = findmnemo(mnemo);

    if (checklabel(label) == FALSE)
    {
      error("malformed label");
    }
  }
  else
  {
    oper = NULL;
    mnemo = NULL;
    label = NULL;
start:

    if (filter(s, "? _?", &mnemo, &oper) && *mnemo && *oper)
    {
      labmnemo = findmnemo(mnemo);

      if (labmnemo == NULL && label == NULL)
      {
        label = mnemo;

        if (checklabel(label) == FALSE)
        {
          error("malformed label");
        }

        s = oper;
        goto start;
      }
    }
    else
    {
      labmnemo = findmnemo(mnemo = s);
      oper = NULL;
    }
  }

  if (labmnemo == NULL)
  {
    if (linestartswithoutblanks && label == NULL && oper == NULL && mnemo != NULL)
    {
      label = mnemo;

      if (status & 2)
      {
        herelabel(label);
      }

      if (status & 1)
      {
        outputline();
      }

      return 0;
    }

    error("unknown macro or mnemonic");
  }

  if (labmnemo->flags & NOLABEL)
  {
    if (label)
    {
      error("labels not allowed here");
    }
  }

  if (labmnemo->flags & DEFLABEL)
  {
    if (!label)
    {
      error("label required here");
    }
  }
  else if (label)
  {
    if (status & 2)
    {
      herelabel(label);
    }
  }

  if (labmnemo->flags & DEFCOND)
  {
    return (*labmnemo->ptr)(status, label, mnemo, oper);
  }

  if (labmnemo->flags & DEFMACRO)
  {
    return (*labmnemo->ptr)(status, label, mnemo, oper);
  }

  if (labmnemo->type == L_MACRO)
  {
    if (label && oper && !strncmp(oper, "MACRO", 5))
    {
      error("macro is already defined");
    }

    if (status & 2)
    {
      macrocall(labmnemo, status, oper);
    }
    else if (status & 1)
    {
      outputline();
    }

    return 0;
  }
  else if (labmnemo->type == L_MNEMO)
  {
    if (status & 2)
    {
      (*labmnemo->ptr)(labmnemo->modifier, label, mnemo, oper);
    }

    starlabel->value += advance;

    if (status & 1)
    {
      outputline();
    }

    return 0;
  }
  else
  {
    fatal("proc asmline failure");
  }

  return 0;
}


/*   F_LIST_ON est pris en compte dans outputline()
     F_CODE_ON est pris en compte dans insertbyte()
     F_NOLIST  est pris en compte dans LIST ON
     F_NOCODE  est pris en compte dans CODE ON ou OFF

     F_NOERROR   dans  error
     F_NOWARNING dans  warning
*/

/*
  ERRORS ROUTINES
*/

void error(char* s)
{
  char raw[80];
  int oldflags;
  char c = toupper(*s);

  oldflags = asmflags;

  if (!(asmflags & F_NOERROR))
  {
    sprintf(raw, "%7d ERROR: %c%s", linenumber, c, s + 1);

    for (s = raw; s[1] == ' '; s++)
    {
      s[0] = '>';
    }

    asmflags |= F_LIST_ON;
    outputraw(raw);
  }

  outputline();

  if (advance != 0)
  {
    starlabel->value += advance;
  }

  asmflags = oldflags;
  longjmp(errorjump, -1);
}


void warning(char* s)
{
  char raw[80];
  char c = toupper(*s);

  if (!(asmflags & F_NOWARNING))
  {
    sprintf(raw, "%7d WARNING: %c%s", linenumber, c, s + 1);

    for (s = raw; s[1] == ' '; s++)
    {
      s[0] = '>';
    }

    outputraw(raw);
  }
}


void fatal(char* s)
{
  printf("FATAL ERROR: %s\n", s);
  exit(10);
}


void fileerror(char* s)
{
  printf("FILE ERROR on %s file: %s\n", s, strerror(errno));
}


/*  aff CPU known  */
extern struct cpulist
{
  char* name;
  int code;
  int (*init)();
} cpulist[];


void syntax(char* s)
{
  register struct cpulist* q;
  register int(*a)();

  a = NULL;
  printf("%s\n", s);
  printf("Syntax:  crasm [-slx] [-o SCODEFILE] INPUTFILE\n");
  printf("Crasm %s known CPUs:", CRASMVERSION);

  for (q = cpulist; q->name ; q++)
  {
    if (a != q->init)
    {
      printf("\n\t");
    }

    printf(" %s", q->name);
    a = q->init;
  }

  printf("\n");
  exit(10);
}


void setflag(int f)
{
  asmflags |= f;
}


void clrflag(int f)
{
  asmflags &= ~f;
}

